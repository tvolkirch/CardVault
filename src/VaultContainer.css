#vault-container
{
    display: flex;
    flex-wrap: wrap;
    justify-content: center;
    margin-top: 3.5rem;
    margin-bottom: 3rem;
}

#top-of-page
{
    visibility: hidden;
    position: fixed;
    bottom: 3.5rem;
    right: 1rem;
    color: #f1f1f1;
    font-weight: bold;
    background-color: #33c;
    margin: 0;
    padding: 0.3rem;
    border: 2px solid #f1f1f1;
}

#top-of-page:hover
{
    cursor: pointer;
    background-color: #36f;
}

#top-of-page:focus
{
    background-color: lime;
}
