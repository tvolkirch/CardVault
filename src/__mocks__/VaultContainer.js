import React from 'react';

/* This is a mocked version of VaultContainer.js */

const VaultContainer = function() { return <div id="vault-container"></div>; };

export default VaultContainer;