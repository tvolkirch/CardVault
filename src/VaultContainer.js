import React, { Component } from "react";
import axios from "axios";

import Loading from "./Loading";
import Card from "./Card";

import "./VaultContainer.css";

class VaultContainer extends Component
{
    _isMounted = false;

    constructor(props, context)
    {
        super(props, context);

        this.state = {
            loading: false,
            cards: [],
            nextApiCall: "https://api.elderscrollslegends.io/v1/cards?page=1&pageSize=20"
        };
        
        // implement infinte scroll with scroll listener and scrollHandler() function to 
        // check for bottom of the scroll window, where it will call the api to load more 
        // cards - the api graciously returns the URL for the next page of cards
        
        var self = this;  // because closure
        var oldPagePosition = 0;
   
        var scrollHandler = function()
        {
            var pageHt = window.innerHeight,
                contentHt = document.body.scrollHeight,
                vScroll = document.body.scrollTop,
                deltaY;
    
            if (vScroll === 0)
            {
                vScroll = document.documentElement.scrollTop;
            }
    
            // set distance from the bottom of the page to trigger data loading
            
            deltaY = contentHt - pageHt - 20;
          
            // scrolled within range of the bottom to trigger another content load
            
            if ( vScroll > deltaY )  
            {
                // help minimize multiple content loadings
                
                if ( vScroll !== oldPagePosition )  
                {
                    self.setState( {loading: true} );
                        
                    setTimeout( function()  
                    {
                        self.callApi();
                    }, 500);
                    
                    oldPagePosition = vScroll;
                }
            }
        };
       
        var scrollTimer;
             
        document.addEventListener("scroll", 
            function() 
            {
                // limit scroll listener to a half second so it doesn't fire too often
                
                if (scrollTimer) 
                {
                    window.clearTimeout(scrollTimer);
                }
             
                scrollTimer = window.setTimeout(scrollHandler, 500);
            }
        );
    }
        
    componentDidMount()
    {
        this._isMounted = true;
    
        this.callApi();
    }
    
    componentWillUnmount()
    {
        this._isMounted = false;
    }
    
    callApi()
    {
        const apiUrl = this.state.nextApiCall;
    
        axios.get(apiUrl)
            .then(response => {
                if (this._isMounted)
                {
                    const cards = this.state.cards.concat(response.data.cards);
                    const nextApiCall = response.data._links.next;
                    this.setState({ loading: false, cards: cards, nextApiCall: nextApiCall });
                }
            })
    }
    
    render()
    {
        var loader = "";
        
        if (this.state.loading)
        {
            loader = <Loading bottom={200} right={150} />;
        }
    
        const cards = this.state.cards.map((card, index) =>
              <Card key={index} 
                 url={card.imageUrl}
                 name={card.name}
                 type={card.type}
                 set={card.set.name}
                 rules={card.text}/>
        );
        
        return (
           <div id="vault-container">{cards}{loader}</div>
        );
    }
}

export default VaultContainer;