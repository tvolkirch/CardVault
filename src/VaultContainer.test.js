import React from "react";
import renderer from "react-test-renderer";

import VaultContainer from "./VaultContainer";

jest.mock("./api/CardService");

it("Consistently renders Vault Container with one mocked card", () =>
{
   const app = renderer
      .create(<VaultContainer />)
      .toJSON();

   expect(app).toMatchSnapshot();
});