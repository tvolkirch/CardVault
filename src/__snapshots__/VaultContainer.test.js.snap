// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consistently renders Vault Container with one mocked card 1`] = `
<div
  id="vault-container"
>
  <div
    className="flex-card-wrapper"
  >
    <div
      className="card-content"
    >
      <img
        alt="Elder Scrolls Card: Mocked Card"
        className="card-image"
        src="#"
      />
    </div>
    <div
      className="card-content"
    >
      <div
        className="card-name"
      >
        Mocked Card
      </div>
      <div>
        <div
          className="flex-card-row"
        >
          <span
            className="left-column"
          >
            Type:
          </span>
          <span
            className="right-column"
          >
            Mocked
          </span>
        </div>
        <div
          className="flex-card-row"
        >
          <span
            className="left-column"
          >
            Set:
          </span>
          <span
            className="right-column"
          >
            Mocked Set
          </span>
        </div>
      </div>
      <div
        className="card-rules"
      >
        This is mocked data for unit testing.
      </div>
    </div>
  </div>
  
</div>
`;
